### ISSUE TITLE
DeepSeek-V4-Flash-0731: deterministic wrong token on deep-context exact retrieval at 1-in-4 prompt lengths (reasoning off) — vLLM 0.28.0 + SGLang, reproduced on two hosted providers (DeepInfra, Baidu); one provider (OpenInference) is correct

### ISSUE THREAD

**Issue by pro-Anne** [vllm-project/vllm#55927]
## Summary
With reasoning disabled, an exact-string needle placed ~28k tokens into a ~57k-token prompt is returned with its last digit wrong (`PROVOCATIVE-8417` → `PROVOCATIVE-8411`) whenever the prompt length ≡ 3 (mod 4) tokens. It is deterministic (temperature 0, same wrong token every time), independent of the surrounding text (fully repetitive filler and a mixed filler both fail), and independent of the top-k backend. With reasoning enabled the same prompts are 16/16 correct. Needles early in the context, or with padding *after* the needle, are always correct.

The same checkpoint served by another provider (OpenRouter → OpenInference, fp8, reasoning_tokens=0) returns the correct code at the same prompt lengths; served by DeepInfra (fp8) and by Baidu (fp8) it returns the same wrong `8411` at every failing length. Four independent stacks agree on the wrong token; one provider does not — so this is not a property of the weights and not specific to our GPU.

## Environment (reporter)
- GPU: 4× NVIDIA RTX PRO 6000 Blackwell Server Edition (sm_120), driver 580.173.02, CUDA 13.0
- vLLM 0.28.0 (`vllm/vllm-openai:v0.28.0`), torch 2.13.0+cu130, flashinfer 0.6.16.post3 (+cubin, +jit-cache cu130), vendored DeepGEMM, nvidia-cutlass-dsl 4.6.2
- Checkpoint: `deepseek-ai/DeepSeek-V4-Flash-0731` @ `7872f01b1d1fe23eabc4c98b48bffcef5a386062` (official; `quantization_config.scale_fmt=ue8m0`)
- Launch (from the official recipe's rtx_pro_6000_8x pill, TP=4):
  `--trust-remote-code --tensor-parallel-size 4 --kv-cache-dtype fp8 --block-size 256 --attention_config.use_fp4_indexer_cache False --tokenizer-mode deepseek_v4 --tool-call-parser deepseek_v4 --enable-auto-tool-choice --reasoning-parser deepseek_v4 --reasoning-config '{"reasoning_parser":"deepseek_v4","reasoning_start_str":"","reasoning_end_str":""}' --max-model-len 262144 --max-num-seqs 64 --gpu-memory-utilization 0.90 --default-chat-template-kwargs '{"thinking": false}'`
- Startup log lines of note: `Selected DeepGemmFp8BlockScaledMMKernel for Fp8LinearMethod`, `Using FP8 indexer cache for Lightning Indexer`, `DeepGEMM E8M0 enabled`, `Using DeepSeek's fp8_ds_mla KV cache format`, `[AutoTuner]: Tuning sparse_mla_sm120_decode_dsv4`
- Also reproduced (identical lengths, identical wrong token) on SGLang v0.5.18 and `nightly-dev-20260901-07c8f729` (`--moe-runner-backend flashinfer_mxfp4 --kv-cache-dtype fp8_e4m3`, dsv4 attention backend), with `--dsa-topk-backend sgl-kernel` and `torch`.

## Minimal reproduction
```python
import json, urllib.request
URL = "http://127.0.0.1:8000/v1/chat/completions"; M = "deepseek-v4-flash-0731"
fox = ("The quick brown fox jumps over the lazy dog. " * 20 + "\n") * 285   # ~57k tokens
code = "PROVOCATIVE-8417"; Q = "\n\nWhat is the secret access code? Answer with just the code."
for n in range(16):  # 16 consecutive prompt lengths; every 4th fails
    q = "pad " * n + fox[:len(fox)//2] + f"\nThe secret access code is {code}.\n" + fox[len(fox)//2:] + Q
    b = {"model": M, "messages": [{"role": "user", "content": q}], "max_tokens": 64, "temperature": 0,
         "chat_template_kwargs": {"thinking": False}}
    d = json.load(urllib.request.urlopen(urllib.request.Request(URL, data=json.dumps(b).encode(), headers={"Content-Type": "application/json"})))
    c = d["choices"][0]["message"]["content"]; print(d["usage"]["prompt_tokens"], "OK" if code in c else f"WRONG {c!r}")
```
Observed (vLLM 0.28.0, thinking off):
```
57316 OK 57317 OK 57318 OK 57319 WRONG 'PROVOCATIVE-8411' 57320 OK 57321 OK 57322 OK 57323 WRONG 'PROVOCATIVE-8411'
57324 OK 57325 OK 57326 OK 57327 WRONG 'PROVOCATIVE-8411' 57328 OK 57329 OK 57330 OK 57331 WRONG 'PROVOCATIVE-8411'
```
Same 16 lengths with `"thinking": True`: 16/16 OK. Mixed (non-repetitive) filler, thinking off: 4/16 wrong at 56730/56734/56738/56742 (same period). Padding appended *after* the needle instead of before it: 16/16 OK. Needle at ~4k tokens instead of ~28k: 16/16 OK.

## Cross-provider control (same checkpoint via OpenRouter, `reasoning: {enabled: false}`, reasoning_tokens=0, identical prompt_tokens)
| prompt_tokens | reporter vLLM 0.28.0 (sm_120) | reporter SGLang (sm_120) | OpenInference (fp8) | DeepInfra (fp8) | Baidu (fp8) |
|---|---|---|---|---|---|
| 57319 (bad) | 8411 | 8411 | refusal (unrelated) | **8411** | **8411** |
| 57323 (bad) | 8411 | 8411 | **8417** ✓ | **8411** | **8411** |
| 57327 (bad) | 8411 | 8411 | **8417** ✓ | **8411** | **8411** |
| 57331 (bad) | 8411 | 8411 | **8417** ✓ | **8411** | **8411** |
| 57318 / 57322 (good) | 8417 ✓ | 8417 ✓ | 8417 ✓ | 8417 ✓ | 8417 ✓ |

Four independent serving stacks (reporter's ×2, DeepInfra, Baidu) agree on the identical wrong token at the identical lengths; one provider (OpenInference) does not. Whether OpenInference deviates from the reference attention path (e.g. dense fallback) or has a fix is unknown to us. DeepSeek's first-party endpoint: not obtainable through our OpenRouter key policy.

## Expected
Exact retrieval should not depend on prompt length mod 4. The failing set is the DSA sparse-attention path with the needle deep in context; the mod-4 period and the single-token corruption suggest an indexer/logits or paged sparse-attention alignment issue rather than a model behavior.

_Repro script and provider-control scripts available on request; the minimal repro above is self-contained._

**Comment by pro-Anne** [comment]
**Update: enabling reasoning is not a workaround — it only lowers the rate and changes the wrong token.**

The DSV4 chat template with `thinking: true` (the default, "high") inserts a 95-token instruction block right after BOS (+79 tokens net vs the thinking-off template; `reasoning_effort: low` encodes identically to thinking-off, `max` adds +92). So a thinking-on sweep shifts every document position by 79 ≡ 3 (mod 4), which is why 16-consecutive-length sweeps with thinking on looked clean — they hit the bad residue with only a ~2% per-case failure rate.

Fine-grained sweep, thinking on (`chat_template_kwargs={"thinking": true}`, temperature 0, max_tokens 2048), same document as above, needle deep, **1-token steps across thinking-off-equivalent lengths 56712..57938 (1,227 cases, pad inserted immediately before the needle)** on the same 4× RTX PRO 6000 / vLLM 0.28.0 setup:

- **6/1227 wrong**, at thinking-on prompt lengths 57140, 57172, 57756, 57936, 57976, 57980 — all ≡ 0 (mod 4).
- Wrong content every time: `PROVOCATIVE-841521` (extra digits), with abnormally long reasoning (100–650 tokens vs a median of 40).
- Re-running the six alone at concurrency 1, twice: **57140 and 57976 fail 2/2 (deterministic)**; the other four pass alone but fail again under concurrency-4 batching (2 and 3 wrong of 24 in two passes), i.e. batch-dependent at the same residue.

So with reasoning off the failure is deterministic and 100% at prompt_len ≡ 3 (mod 4) (`8411`); with reasoning on it is ~2% at the shifted residue, partly deterministic, partly batch-sensitive (`841521`). Both point at the same alignment-dependent path. Raw requests/responses, the exact encoded prompts (token ids) for a failing and a passing case, and the sweep harness are available on request.

**Comment by debanshd** [comment]
I am looking into this retrieval corruption. The deterministic failure at `prompt_length == 3 (mod 4)` points to an unaligned 4-token stride boundary in the rotary embedding and context chunking offset calculation. I would like to help investigate and isolate the index offset fix for this.

**Comment by ZC502** [comment]
@pro-Anne The updated sweep makes this especially interesting as an **input-geometry** problem rather than just a single failing prompt: the residue stays structured, while the failure changes from fully deterministic to partly batch-sensitive once the template shifts the token geometry.
I’ve been working on a small measurement-only position-parity format for vLLM. Your case exposes a useful gap in it: the current analyzer assumes aligned prompt tokenization across arms, which is appropriate for same-input backend/config comparisons, but not for a one-token-at-a-time geometry sweep like this.

A geometry-aware extension would need to treat things like:
- prompt token count / residue
- needle token offset
- template offset
- concurrency / batch condition
- passing vs failing neighboring lengths
- observed output/logprob differences where available

as explicit experimental coordinates, rather than pretending `position N` in two different-length prompts is automatically the same semantic position.
You mentioned that the exact encoded prompts/token IDs and sweep harness are available. If you’re willing to share one adjacent passing/failing pair plus the harness/output shape, I’d be happy to prototype that postprocessor against your existing data first.
I would keep it measurement-only — no claim that a particular rotary/indexer/attention kernel is causal. If the representation proves useful here, it could become an **input-geometry parity** extension of the same canonical evidence schema rather than a one-off DeepSeek script.

Current measurement tool for context: vllm-position-parity — [https://github.com/ZC502/vllm-position-parity.git](https://github.com/ZC502/vllm-position-parity.git)
